### PR TITLE
copyKAT function

### PR DESCRIPTION
**Pull request by abyssum**

### CODE REVIEW

**Comment by abyssum** [comment]
Needs rpy2 and an active R installation (with pkg "copykat" installed):
```
library(devtools)
install_github("navinlabcode/copykat")
```

File: infercnvpy/tl/_copykat.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    subclones using high-throughput sc-RNAseq data.
+    
+    Note: The matrix values are often the count of unique molecular identifier (UMI) 
+    from nowadays high througput single cell RNAseq data. The early generation of 
+    scRNAseq data may be summarized as TPM values or total read counts, 
+    which should also work.
+
+    You can find more info on GitHub: https://github.com/navinlabcode/copykat
+
+    DOI: https://doi.org/10.1038/s41587-020-00795-2
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:29]: Could you please add this to `docs/references.bib`? 

Then you can do `` :cite:`WhateverBibtexId` `` somewhere in the text and it gets formatted as citation.

File: infercnvpy/tl/_copykat.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import infercnvpy as cnv
+import numpy as np
+import pandas as pd
+import scipy
+from scipy.sparse import issparse
+from anndata import AnnData
+from scanpy import logging, AnnData
+
+def copyKAT(
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:9]:
```suggestion
def copykat(
```

let's use lowercase everywhere to be consistent with python code-style.

Comment (grst) [inline infercnvpy/tl/_copykat.py:9]: `copykat` neets to be exported to `infercnvpy/tl/__init__.py` to be accessible through the API.

File: infercnvpy/tl/_copykat.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        ro.default_converter + numpy2ri.converter + pandas2ri.converter
+    ):
+        copyKAT_result = ro.conversion.rpy2py(ro.globalenv["copyKAT_result"])
+
+    adata.uns["copyKAT"] = {
+        "chr_pos": {
+            f"chr{chrom}": int(pos)
+            for pos, chrom in copyKAT_result.loc[:, ["chrom"]].drop_duplicates().itertuples()
+        }
+    }
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:95]: Please store this in a temporary variable and only assign it to `adata` if `inplace == True`

File: infercnvpy/tl/_copykat.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        colnames(copyKAT_result) <- c('chrom', 'chrompos', 'abspos', cell_IDs)
+        """
+    )
+
+    with localconverter(
+        ro.default_converter + numpy2ri.converter + pandas2ri.converter
+    ):
+        copyKAT_result = ro.conversion.rpy2py(ro.globalenv["copyKAT_result"])
+
+    adata.uns["copyKAT"] = {
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:90]:
```suggestion
    adata.uns[key_added] = {
```

File: infercnvpy/tl/_copykat.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import infercnvpy as cnv
+import numpy as np
+import pandas as pd
+import scipy
+from scipy.sparse import issparse
+from anndata import AnnData
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:6]: duplicate import of anndata. Doesnt matter which, they point to the same object in the end.

File: infercnvpy/tl/_copykat.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    You can find more info on GitHub: https://github.com/navinlabcode/copykat
+
+    DOI: https://doi.org/10.1038/s41587-020-00795-2
+
+    Parameters
+    ----------
+    adata
+        annotated data matrix
+    key_added
+        Key under which the copyKAT scores will be stored in `adata.obs`.
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:36]:
```suggestion
        Key under which the copyKAT scores will be stored in `adata.obsm` and `adata.uns`.
```

**Review by grst — CHANGES_REQUESTED** [review]

**Comment by grst** [comment]
Other TODOs (but doesn't need to be in this PR, more a note for myself): 

 - [ ] Add entry to API documentation
 - [ ] Update Installation instructions
 - [ ] Add example to tutorial
  - [ ]  Add unittest for copykat
  - [ ]  Update CI to be able to run R packages
  - [ ]  Ideally, add copykat to bioconda

File: infercnvpy/tl/_copykat.py (hunk 1 above)
Comment (abyssum) [reply infercnvpy/tl/_copykat.py]: done

File: infercnvpy/tl/_copykat.py (hunk 2 above)
Comment (abyssum) [reply infercnvpy/tl/_copykat.py:9]: on it

File: infercnvpy/tl/_copykat.py (hunk 5 above)
Comment (abyssum) [reply infercnvpy/tl/_copykat.py:6]: my bad

File: infercnvpy/tl/_copykat.py (hunk 2 above)
Comment (abyssum) [reply infercnvpy/tl/_copykat.py]: done

File: infercnvpy/tl/_copykat.py (hunk 3 above)
Comment (abyssum) [reply infercnvpy/tl/_copykat.py]: and done

File: infercnvpy/tl/_copykat.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    ro.globalenv["cell_IDs"] = ro.conversion.py2rpy(list(adata.obs.index))
+    
+    logging.info("Running copyKAT")
+    ro.r(
+        """
+        rownames(expr_r) <- gene_names
+        colnames(expr_r) <- cell_IDs
+        copyKAT_run <- copykat(rawmat = expr_r, id.type = "S", ngene.chr = 5, win.size = 25, 
+                                KS.cut = 0.1, sam.name = "test", distance = "euclidean", norm.cell.names = "", 
+                                n.cores = 12, output.seg = FALSE)
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:77]: @abyssum, do you think it's worth exposing these parameters to the python user?

Comment (abyssum) [reply infercnvpy/tl/_copykat.py]: yes, especially cores, distance, and KS.cut... I was thinking about it as well, on it!

Comment (abyssum) [reply infercnvpy/tl/_copykat.py]: done

File: infercnvpy/tl/_copykat.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+
+def copykat(
+    adata: AnnData,
+    gene_ids: str = "S",
+    segmentation_cut: float = 0.1,
+    distance: str = "euclidean",
+    s_name: str = "copykat_result",
+    min_genes_chr: int = 5,
+    key_added: str = "copyKAT",
```
Comment (grst) [inline infercnvpy/tl/_copykat.py:17]:
```suggestion
    key_added: str = "cnv",
```

**Review by grst — APPROVED** [review]
LGTM! Thanks a lot, @abyssum

**Pull request merged**